mailserver/mailserver.go

```diff
@@ -247,25 +250,66 @@ func (s *WMailServer) DeliverMail(peer *whisper.Peer, request *whisper.Envelope)
 	}
 	if s.exceedsPeerRequests(peer.ID()) {
 		requestErrorsCounter.Inc(1)
+		log.Error("Peer requests limit exceeded")
 		return
 	}
 
 	defer recoverLevelDBPanics("DeliverMail")
 
-	if ok, lower, upper, bloom, limit, cursor := s.validateRequest(peer.ID(), request); ok {
-		_, lastEnvelopeHash, nextPageCursor, err := s.processRequest(peer, lower, upper, bloom, limit, cursor)
-		if err != nil {
-			processRequestErrorsCounter.Inc(1)
-			log.Error(fmt.Sprintf("error in DeliverMail: %s", err))
-			return
-		}
+	var (
+		ok           bool
+		lower, upper uint32
+		bloom        []byte
+		limit        uint32
+		cursor       cursorType
+		batch        bool
+	)
 
-		if err := s.sendHistoricMessageResponse(peer, request, lastEnvelopeHash, nextPageCursor); err != nil {
-			historicResponseErrorsCounter.Inc(1)
-			log.Error(fmt.Sprintf("SendHistoricMessageResponse error: %s", err))
-		}
+	if payload, err := s.decodeRequest(peer.ID(), request); err == nil {
+		ok = true
+		lower, upper = payload.Lower, payload.Upper
+		bloom = payload.Bloom
+		cursor = cursorType(payload.Cursor)
+		limit = payload.Limit
+		batch = payload.Batch
 	} else {
+		log.Debug("Failed to decode request", "err", err, "peerID", peer.ID())
+		log.Debug("Fallback to old request payload")
+		ok, lower, upper, bloom, limit, cursor = s.validateRequest(peer.ID(), request)
+	}
+
+	if !ok {
 		requestValidationErrorsCounter.Inc(1)
+		log.Error("Failed to validate request")
+		return
+	}
+
+	log.Debug("Processing request",
+		"lower", lower, "upper", upper,
+		"bloom", bloom,
+		"limit", limit,
+		"cursor", cursor,
+		"batch", batch)
+
+	_, lastEnvelopeHash, nextPageCursor, err := s.processRequest(
+		peer,
+		lower, upper,
+		bloom,
+		limit,
+		cursor,
+		batch)
+	if err != nil {
+		processRequestErrorsCounter.Inc(1)
+		log.Error("Error while processing mail server request", "err", err)
+		return
+	}
+
+	log.Debug("Processed request successfully")
+	log.Debug("Sending historic message response", "last", lastEnvelopeHash, "next", nextPageCursor)
```

Comment:
two logs are intentional?

Reply:
Yes but the first one is not needed. I will remove it.